src/Stack/Setup.hs

```diff
@@ -489,11 +489,12 @@ ensureCompiler sopts = do
                 m <- augmentPathMap (edBins ed) (unEnvOverride menv0)
                 mkEnvOverride (configPlatform config) (removeHaskellEnvVars m)
 
-    when (soptsUpgradeCabal sopts) $ do
+    let cabalVersion = soptsUpgradeCabal sopts
+    when (isJust $ cabalVersion) $ do
         unless needLocal $ do
-            $logWarn "Trying to upgrade Cabal library on a GHC not installed by stack."
+            $logWarn "Trying to change a Cabal library on a GHC not installed by stack."
             $logWarn "This may fail, caveat emptor!"
-        upgradeCabal menv wc
+        upgradeCabal menv wc (fromJust cabalVersion)
```

Comment:
Not sure how to get the `when isJust` out of it?. See https://github.com/commercialhaskell/stack/pull/2995/commits/734dae44615208d8404229cf1fc64e7ec989bbf7.